log4j-api/src/main/java/org/apache/logging/log4j/util/ServiceLoaderUtil.java

```diff
@@ -110,28 +139,64 @@ private ServiceLoaderUtil() {
      * @param logger      An action to perform for each broken service.
      * @param services    A list to add the services.
      */
-    private static <T> void addServices(final Class<T> serviceType, final ClassLoader classLoader,
-            final Consumer<Throwable> logger, final Collection<T> services) {
-        final Iterator<T> iterator = ServiceLoader.load(serviceType, classLoader).iterator();
-        while (iterator.hasNext()) {
-            try {
-                final T service = iterator.next();
-                if (classLoader.equals(service.getClass().getClassLoader())) {
-                    services.add(service);
-                }
-            } catch (ServiceConfigurationError e) {
-                if (logger != null) {
-                    logger.accept(e);
-                }
-            }
-        }
+    private static <T> Stream<T> loadClassloaderServices(final Class<T> serviceType, final ClassLoader classLoader,
+            final boolean verbose) {
+        return StreamSupport.stream(new ServiceLoaderSpliterator<T>(serviceType, classLoader, verbose), false);
     }
 
-    private static <T> void addOsgiServices(final Class<T> serviceType, final Collection<T> services) {
+    private static <T> Stream<? extends T> loadOsgiServices(final Class<T> serviceType) {
         final Iterable<? extends T> iterable = org.glassfish.hk2.osgiresourcelocator.ServiceLoader
                 .lookupProviderInstances(serviceType);
-        if (iterable != null) {
-            iterable.forEach(services::add);
+        return iterable != null ? StreamSupport.stream(iterable.spliterator(), false) : Stream.empty();
+    }
+
+    private static class ServiceLoaderSpliterator<S> implements Spliterator<S> {
+
+        private final Iterator<S> serviceIterator;
+        private final ClassLoader classLoader;
+        private final Logger logger;
+        private final String serviceName;
+
+        public ServiceLoaderSpliterator(final Class<S> serviceType, final ClassLoader classLoader,
+                final boolean verbose) {
+            this.serviceIterator = ServiceLoader.load(serviceType, classLoader).iterator();
```

Comment:
This code will not work for applications using the module path. Currently, all calls to ServiceLoader happen in the jar that needs the service. However, according to https://openjdk.java.net/projects/jigsaw/spec/sotms/#services, if an automatic module declares a service it behaves as if it had a corresponding provides clause.  Note that log4j-api has a module-info.java, so it will act as an explicitly named module. Currently, it declares uses clauses for the services log4j-api needs. This change will mean it will require a uses clauses for any use of ServiceLoader, which makes no sense and certainly won't work as expected.

Reply:
@rgoers,
Thanks for the explanations.

In `master` you use a lambda that calls [`ServiceLoader#load(ModuleLayer, Class)`](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/ServiceLoader.html#load(java.lang.ModuleLayer,java.lang.Class)) to work around the caller semantics. This obviously can not work in 2.x, since `ModuleLayer` does not exist and we would need a Java9 version of each class that uses the `ServiceLoaderUtil`. Do you know any tricks that can be applied to Java 8? Would adding a `MethodHandles.Lookup` parameter work?

If it works, we could have a

```lang-java
ServiceLoaderUtil#loadServices(Class<T> serviceType, MethodHandles.Lookup lookup, boolean useTccl)
```
signature and look for services in the caller's classloader and the thread context classloader (if `useTccl=true`).

@jvz, any suggestions?

Reply:
@ppkarwasz No matter how you slice it, so long as the call originates from the log4j-api module it won't work in JPMS. I would suggest that you use ServiceLoader.load(Class<S> service, ClassLoader classLoader) from the lambda. You will end up calling it once for each ClassLoader you have. In master I don't have to do this since anything running on the module path is going to have a ModuleLayer. 
Another option, of course, is to create a Java 9 specific version of this that supports using ModuleLayer. Then the code you currently have would work. That is probably the route I would go.